File: backend/app/schemas/notifications.py
"""Notification schemas (PHASE 13).

Contracts for the customer notification centre and the admin notification desk.
Rules that shape every model here:

  * A notification carries REFERENCES and STATUSES only — never a card detail,
    OTP, password, passport/document number or provider credential.
  * Nothing a client sends decides audience, ownership or admin level: those are
    resolved from the verified bearer token server-side.
  * A delivery state describes what OUR service attempted. No real Email/SMS/
    WhatsApp provider is connected, so demo attempts are labelled as such.
"""

from __future__ import annotations

from datetime import datetime
from typing import Literal, Optional

from pydantic import BaseModel, ConfigDict, Field

NotificationChannel = Literal["in_app", "email", "sms", "whatsapp", "push"]
NotificationDeliveryState = Literal["queued", "sent", "failed", "retrying", "skipped"]
NotificationAudience = Literal["customer", "admin"]

NotificationType = Literal[
    "booking_created",
    "booking_pending",
    "booking_confirmed",
    "booking_failed",
    "booking_cancelled",
    "booking_update",
    "payment_successful",
    "payment_failed",
    "payment_refunded",
    "cancellation_requested",
    "cancellation_status_changed",
    "refund_status_changed",
    "documents_available",
    "support_request_created",
    "support_request_replied",
    "support_request_updated",
]


class Model(BaseModel):
    model_config = ConfigDict(extra="forbid", populate_by_name=True)


class NotificationDeliveryOut(Model):
    channel: NotificationChannel
    state: NotificationDeliveryState
    attempts: int = 0
    provider_id: str = Field(default="none", serialization_alias="providerId")
    mode: Literal["demo", "live"] = "demo"
    updated_at: datetime = Field(serialization_alias="updatedAt")
    error: Optional[str] = None


class NotificationOut(Model):
    """Camel-case aliases keep the frontend `NotificationRecord` shape."""

    model_config = ConfigDict(extra="forbid", populate_by_name=True, serialize_by_alias=True)

    id: str
    type: NotificationType
    title: str
    body: str
    created_at: datetime = Field(serialization_alias="createdAt")
    booking_reference: Optional[str] = Field(default=None, serialization_alias="bookingReference")
    support_request_id: Optional[str] = Field(default=None, serialization_alias="supportRequestId")
    audience: NotificationAudience = "customer"
    channels: list[NotificationChannel] = Field(default_factory=list)
    dispatched: bool = False
    read_at: Optional[datetime] = Field(default=None, serialization_alias="readAt")
    deliveries: list[NotificationDeliveryOut] = Field(default_factory=list)


class UnreadCountOut(Model):
    count: int = 0


class NotificationActionOut(Model):
    ok: bool = True
    """Always true when the request was accepted. Delivery is asynchronous and
    never blocks a booking or a payment."""

File: src/types/notifications.ts
/**
 * PHASE 13 — notification contracts.
 *
 * These types describe the delivery layer that sits on top of the Phase 11
 * `notification_events` records. Rules that hold everywhere in this file:
 *
 *  - a notification carries REFERENCES and STATUSES only. Never a card detail,
 *    OTP, password, passport/document number, provider credential or a full
 *    contact record,
 *  - a delivery state describes what our own service attempted. It never
 *    claims a provider confirmed anything, because no real Email/SMS/WhatsApp
 *    provider is connected yet,
 *  - delivery NEVER decides booking or payment success. A failed notification
 *    is a notification problem, nothing else.
 */

import type { NotificationEvent, NotificationType } from "@/types/operations";

export type NotificationChannel = "in_app" | "email" | "sms" | "whatsapp" | "push";

/** Safe delivery states. `skipped` = no provider is wired for that channel. */
export type NotificationDeliveryState = "queued" | "sent" | "failed" | "retrying" | "skipped";

export interface NotificationDelivery {
  channel: NotificationChannel;
  state: NotificationDeliveryState;
  /** How many send attempts the service has made (never unbounded). */
  attempts: number;
  /** Provider identifier, e.g. "demo-email". Never a key or endpoint. */
  providerId: string;
  /** Provider mode, so the UI can label demo delivery honestly. */
  mode: "demo" | "live";
  updatedAt: string;
  /** Short, non-sensitive failure reason for the admin desk. */
  error?: string | null;
}

export type NotificationAudience = NotificationEvent["audience"];

/** A stored notification plus its read state and delivery attempts. */
export interface NotificationRecord extends NotificationEvent {
  /** Null while unread. */
  readAt?: string | null;
  /** Idempotency key — a repeat emit with the same key is ignored. */
  dedupeKey?: string | null;
  deliveries?: NotificationDelivery[];
}

/** What the centralized service is asked to emit. */
export interface NotificationEmitInput {
  type: NotificationType;
  /** Optional override; the template supplies a default. */
  title?: string;
  body?: string;
  audience?: NotificationAudience;
  bookingReference?: string | null;
  supportRequestId?: string | null;
  /** Restrict the channels for this one event (defaults to the template). */
  channels?: NotificationChannel[];
  /** Explicit idempotency key. Defaults to type + reference + title. */
  dedupeKey?: string | null;
}

/** The payload a channel provider receives. Contains no PII beyond references. */
export interface NotificationMessage {
  notificationId: string;
  channel: NotificationChannel;
  type: NotificationType;
  title: string;
  body: string;
  bookingReference?: string | null;
  supportRequestId?: string | null;
}

export interface NotificationSendResult {
  ok: boolean;
  /** Short reason when `ok` is false. Never a stack trace or credential. */
  error?: string;
}

/**
 * Channel provider abstraction.
 *
 * Booking, payment and support code never talks to a provider: it emits an
 * event through the notification service, and the service picks the provider
 * registered for each channel. Swapping the demo provider for a real one is a
 * registry change, nothing more.
 */
export interface NotificationProvider {
  id: string;
  channel: NotificationChannel;
  mode: "demo" | "live";
  send(message: NotificationMessage): Promise<NotificationSendResult>;
}

export interface NotificationListQuery {
  audience?: NotificationAudience;
  unreadOnly?: boolean;
  bookingReference?: string | null;
  limit?: number;
}
